### ISSUE TITLE
some problems about multi gpu training

### ISSUE THREAD

**Issue by frost-768** [HaoyuHu/bert-multi-gpu#8]
Thanks a lot for providing the multi-gpu-support for bert. I have some problems when running your bert-multi-gpu code "run_custom_classifier.py ".

1. line 929. log_every_n_steps = 8
   When i use 2 gpus the log prints every 16 step. However when i use 4 gpus, the log still prints every 16 steps. Should it be 8*4 steps?
In addition, the running time for 16 steps is almost the same no matter using 2 or 4 gpus (4 gpus evern cost more time). So what does the log 16 steps meam? Does it mean each gpu will run a different 16 steps?

2. line 953. num_train_steps = int(len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs)
If I use 4 gpus, should num_train_steps = len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs / 4 ?

**Comment by haoyuhu** [comment]
> Thanks a lot for providing the multi-gpu-support for bert. I have some problems when running your bert-multi-gpu code "run_custom_classifier.py ".
> 
> 1. line 929. log_every_n_steps = 8
>    When i use 2 gpus the log prints every 16 step. However when i use 4 gpus, the log still prints every 16 steps. Should it be 8*4 steps?
>    In addition, the running time for 16 steps is almost the same no matter using 2 or 4 gpus (4 gpus evern cost more time). So what does the log 16 steps meam? Does it mean each gpu will run a different 16 steps?
> 2. line 953. num_train_steps = int(len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs)
>    If I use 4 gpus, should num_train_steps = len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs / 4 ?

1. I think `log_every_n_steps` is for every GPU, so every replica will print log every `log_every_n_steps`. But I'm not sure why it is 16 steps in your use case.
2. `num_train_steps` is for all GPUs, so there is no need to divide by the number of GPUs.

**Comment by wushuangzhi** [comment]
> > Thanks a lot for providing the multi-gpu-support for bert. I have some problems when running your bert-multi-gpu code "run_custom_classifier.py ".
> > 
> > 1. line 929. log_every_n_steps = 8
> >    When i use 2 gpus the log prints every 16 step. However when i use 4 gpus, the log still prints every 16 steps. Should it be 8*4 steps?
> >    In addition, the running time for 16 steps is almost the same no matter using 2 or 4 gpus (4 gpus evern cost more time). So what does the log 16 steps meam? Does it mean each gpu will run a different 16 steps?
> > 2. line 953. num_train_steps = int(len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs)
> >    If I use 4 gpus, should num_train_steps = len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs / 4 ?
> 
> 1. I think `log_every_n_steps` is for every GPU, so every replica will print log every `log_every_n_steps`. But I'm not sure why it is 16 steps in your use case.
> 2. `num_train_steps` is for all GPUs, so there is no need to divide by the number of GPUs.

For num_train_steps, do you mean in training, each gpu will trun num_train_steps ?

**Comment by frost-768** [comment]
> > Thanks a lot for providing the multi-gpu-support for bert. I have some problems when running your bert-multi-gpu code "run_custom_classifier.py ".
> > 
> > 1. line 929. log_every_n_steps = 8
> >    When i use 2 gpus the log prints every 16 step. However when i use 4 gpus, the log still prints every 16 steps. Should it be 8*4 steps?
> >    In addition, the running time for 16 steps is almost the same no matter using 2 or 4 gpus (4 gpus evern cost more time). So what does the log 16 steps meam? Does it mean each gpu will run a different 16 steps?
> > 2. line 953. num_train_steps = int(len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs)
> >    If I use 4 gpus, should num_train_steps = len(train_examples) / FLAGS.train_batch_size * FLAGS.num_train_epochs / 4 ?
> 
> 1. I think `log_every_n_steps` is for every GPU, so every replica will print log every `log_every_n_steps`. But I'm not sure why it is 16 steps in your use case.
> 2. `num_train_steps` is for all GPUs, so there is no need to divide by the number of GPUs.

For num_train_steps, do you mean in training, each gpu will run num_train_steps ?

**Comment by haoyuhu** [comment]
Regardless of the number of GPUs, `num_train_steps` will not change. The `batch_size` here is for each GPU.

**Comment by gzn555823** [comment]
> Regardless of the number of GPUs, `num_train_steps` will not change. The `batch_size` here is for each GPU.

Do you mean in this multi-gpu training, each gpu processes a same batch of data? i.e. with gpu_num=4, it will train 4 times data  as gpu_num=1?

**Comment by gzn555823** [comment]
I found the same problem, the log always prints every `log_every_n_steps*2` steps, even when I use only one gpu.
I also found the training speed of multi-gpu is slower than single gpu. I changed every `int(FLAGS.num_gpu_cores) >= 2` to `int(FLAGS.num_gpu_cores) >= 1` in code, and run a MRPC task, the training speed is e.g.
global_step/sec: 1.37571 (gpu_num=1)
global_step/sec: 1.27179 (gpu_num=2)
but the loss decrease faster with multi-gpu, e.g.: (I have tried many times)
loss = 0.5023794, step = 120 (gpu_num=1)
loss = 0.3498465, step = 120 (gpu_num=2)
Therefore, I think in this training strategy, the num of trained examples in one step is gpu_num*batch_size ( Is the batch distributed to each gpu the same?)

Hope to hear your opinion : )

Addition:
Set use_gpu=false, and use original TPUEstimator(should fall back to Estimator, but MirroredStrategy is not used compared above). The speed is even slower:
global_step/sec: 0.683546
of course GPU memory is used.
Any explanation is appreciated.

**Comment by haoyuhu** [comment]
> Do you mean in this multi-gpu training, each gpu processes a same batch of data? i.e. with gpu_num=4, it will train 4 times data as gpu_num=1?

No, it trains every sample only once in every epoch.

> Therefore, I think in this training strategy, the num of trained examples in one step is gpu_num*batch_size ( Is the batch distributed to each gpu the same?)

You are right. In `MirroredStrategy`, `global_batch_size = num_gpu_cores * batch_size_per_gpu`. So in your use case, the loss decreased faster than ONE GPU.

**Comment by wushuangzhi** [comment]
> I found the same problem, the log always prints every `log_every_n_steps*2` steps, even when I use only one gpu.
> I also found the training speed of multi-gpu is slower than single gpu. I changed every `int(FLAGS.num_gpu_cores) >= 2` to `int(FLAGS.num_gpu_cores) >= 1` in code, and run a MRPC task, the training speed is e.g.
> global_step/sec: 1.37571 (gpu_num=1)
> global_step/sec: 1.27179 (gpu_num=2)
> but the loss decrease faster with multi-gpu, e.g.: (I have tried many times)
> loss = 0.5023794, step = 120 (gpu_num=1)
> loss = 0.3498465, step = 120 (gpu_num=2)
> Therefore, I think in this training strategy, the num of trained examples in one step is gpu_num*batch_size ( Is the batch distributed to each gpu the same?)
> 
> Hope to hear your opinion : )
> 
> Addition:
> Set use_gpu=false, and use original TPUEstimator(should fall back to Estimator, but MirroredStrategy is not used compared above). The speed is even slower:
> global_step/sec: 0.683546
> of course GPU memory is used.
> Any explanation is appreciated.

**Comment by zhuolinumd** [comment]
@HaoyuHu @frost-768 @gzn555823 

> You are right. In `MirroredStrategy`, `global_batch_size = num_gpu_cores * batch_size_per_gpu`. So in your use case, the loss decreased faster than ONE GPU.

Is `global_batch_size` controlling the total number of iterations during optimization?

 Based on the code (https://github.com/HaoyuHu/bert-multi-gpu/blob/089515689f44b6f3e30a1bdae2476336d325942c/run_custom_classifier.py#L953-L954) 

Does it mean that the number of iterations for each GPU is `num_train_steps `? My understanding that with more GPUs, we should be able to train faster? It seems not this case now. Thanks.

**Comment by haoyuhu** [comment]
> Is `global_batch_size` controlling the total number of iterations during optimization?

`FLAGS.train_batch_size`: Batch size for each GPU. For example, if train_batch_size is 16, and num_gpu_cores is 4, your GLOBAL batch size is 16 * 4 = 64.
`train_examples`: Total example number.
So `num_train_steps` is independent of the number of GPU cores.

> Does it mean that the number of iterations for each GPU is `num_train_steps `?

No. The number of iterations is for whole task, not each GPU.

**Comment by zhuolinumd** [comment]
@HaoyuHu Thanks for your reply.

> > Does it mean that the number of iterations for each GPU is `num_train_steps `?
> 
> No. The number of iterations is for whole task, not each GPU.

Do you mean `num_train_steps` is fixed for the whole task, with more number of gpu `num_gpu_cores` can speed up the training? Is this because the data are distributed to more GPUs?

**Comment by haoyuhu** [comment]
> @HaoyuHu Thanks for your reply.
> 
> > > Does it mean that the number of iterations for each GPU is `num_train_steps `?
> > 
> > 
> > No. The number of iterations is for whole task, not each GPU.
> 
> Do you mean `num_train_steps` is fixed for the whole task, with more number of gpu `num_gpu_cores` can speed up the training? Is this because the data are distributed to more GPUs?

Yes. You are right.
